test/integration/apiserver/flowcontrol/concurrency_test.go

```diff
@@ -0,0 +1,249 @@
+
+	streamRequests(nil, 1, 100, func() {
+		_, err := noxu2Client.CoreV1().Namespaces().List(context.TODO(), metav1.ListOptions{})
+		require.NoError(t, err)
+	})
+
+	wg.Wait()
+
+	dispatchedCountNoxu1, err := getRequestCountOfPriorityLevel(loopbackClient, priorityLevelNoxu1.Name)
+	require.NoError(t, err)
+	dispatchedCountNoxu2, err := getRequestCountOfPriorityLevel(loopbackClient, priorityLevelNoxu2.Name)
+	require.NoError(t, err)
+
+	assert.Equal(t, 1000, dispatchedCountNoxu1)
+	assert.Equal(t, 100, dispatchedCountNoxu2)
+}
+
+func getClientFor(loopbackConfig *rest.Config, username string) clientset.Interface {
+	config := &rest.Config{
+		Host:        loopbackConfig.Host,
+		QPS:         -1,
+		BearerToken: loopbackConfig.BearerToken,
+		Impersonate: rest.ImpersonationConfig{
+			UserName: username,
+		},
+	}
+	return clientset.NewForConfigOrDie(config)
+}
+
+func getRequestCountOfPriorityLevel(c clientset.Interface, priorityLevelName string) (int, error) {
+	resp, err := c.CoreV1().
+		RESTClient().
+		Get().
+		RequestURI("/metrics").
+		DoRaw(context.TODO())
+	if err != nil {
+		return 0, err
+	}
+
+	dec := expfmt.NewDecoder(strings.NewReader(string(resp)), expfmt.FmtText)
+	decoder := expfmt.SampleDecoder{
+		Dec:  dec,
+		Opts: &expfmt.DecodeOptions{},
+	}
+
+	for {
+		var v model.Vector
+		if err := decoder.Decode(&v); err != nil {
+			if err == io.EOF {
+				// Expected loop termination condition.
+				return 0, fmt.Errorf("no dispatched-count metrics found for priorityLevel %v", priorityLevelName)
+			}
+			return 0, fmt.Errorf("failed decoding metrics: %v", err)
+		}
+		for _, metric := range v {
+			switch name := string(metric.Metric[model.MetricNameLabel]); name {
+			case dispatchedRequestCountMetricsName:
+				if priorityLevelName == string(metric.Metric[dispatchedRequestCountMetricsLabelPriorityLevel]) {
+					return int(metric.Value), nil
+				}
+			}
+		}
+	}
+}
+
+func createPriorityLevelAndBindingFlowSchemaForUser(c clientset.Interface, username string) (*flowcontrolv1alpha1.PriorityLevelConfiguration, *flowcontrolv1alpha1.FlowSchema, error) {
+	pl, err := c.FlowcontrolV1alpha1().PriorityLevelConfigurations().Create(context.TODO(), &flowcontrolv1alpha1.PriorityLevelConfiguration{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: username,
+		},
+		Spec: flowcontrolv1alpha1.PriorityLevelConfigurationSpec{
+			Type: flowcontrolv1alpha1.PriorityLevelEnablementLimited,
+			Limited: &flowcontrolv1alpha1.LimitedPriorityLevelConfiguration{
+				AssuredConcurrencyShares: 10,
+				LimitResponse: flowcontrolv1alpha1.LimitResponse{
+					Type: flowcontrolv1alpha1.LimitResponseTypeQueue,
+					Queuing: &flowcontrolv1alpha1.QueuingConfiguration{
+						Queues:           100,
+						HandSize:         1,
+						QueueLengthLimit: 10,
+					},
+				},
+			},
+		},
+	}, metav1.CreateOptions{})
+	if err != nil {
+		return nil, nil, err
+	}
+	fs, err := c.FlowcontrolV1alpha1().FlowSchemas().Create(context.TODO(), &flowcontrolv1alpha1.FlowSchema{
```

Comment:
We should really get into the habit of writing objects in YAML and decoding them. Not asking for a change, since we do it wrong everywhere. But YAML is easier to copy/paste and reuse in other contexts.

Reply:
we've been writing tests in this way for a loooong time. and yes writing in YAML-then-decode will help those curious users trying to understand how it works by reading the tests. YAML is easier to copy/paste and verify in their own cluster.